packages/itwin/tree-widget/src/tree-widget-react/components/trees/common/components/Tree.tsx

```diff
@@ -71,26 +71,39 @@ export type TreeProps = Pick<FunctionProps<typeof useIModelTree>, "getFilteredPa
  * @beta
  */
 export function Tree({ getSchemaContext, hierarchyLevelSizeLimit, selectionStorage, imodelAccess: providedIModelAccess, ...props }: TreeProps) {
-  const defaultHierarchyLevelSizeLimit = hierarchyLevelSizeLimit ?? 1000;
-
-  const imodelAccess = useMemo(() => {
-    return providedIModelAccess ?? createIModelAccess({ getSchemaContext, imodel: props.imodel });
-  }, [providedIModelAccess, getSchemaContext, props.imodel]);
+  const { imodelAccess, currentHierarchyLevelSizeLimit } = useIModelAccess({
+    imodel: props.imodel,
+    imodelAccess: providedIModelAccess,
+    getSchemaContext,
+    hierarchyLevelSizeLimit,
+  });
+  return (
+    <TreeBase
+      {...props}
+      getSchemaContext={getSchemaContext}
+      selectionStorage={selectionStorage}
+      imodelAccess={imodelAccess}
+      currentHierarchyLevelSizeLimit={currentHierarchyLevelSizeLimit}
+    />
+  );
+}
 
+/** @internal */
+export function TreeBase({ getSchemaContext, ...props }: MarkRequired<TreeProps, "imodelAccess"> & { currentHierarchyLevelSizeLimit: number }) {
```

Comment:
do we need this? It looks like we could just rename `TreeBaseImpl` to `TreeBase` and export that.

Reply:
It add `SchemaMetadataContextProvider`. Moving it inside `TreeBaseImpl` didn't work. I guess some hook depends on that context. Others solution would be to add `SchemaMetadataContextProvider` in both `Tree` and `VisbilityTree` components. That way we could remove this one.

Reply:
nah, if some hook used in `TreeBaseImpl` needs this context, then I think let's keep this as is